specs/core/0_fork-choice.md

```diff
@@ -136,17 +136,69 @@ def get_latest_attesting_balance(store: Store, root: Root) -> Gwei:
     ))
 ```
 
+#### `filter_block_tree`
+
+```python
+def filter_block_tree(store: Store, block_root: Root, blocks: Dict[Root, BeaconBlock]) -> bool:
+    block = store.blocks[block_root]
+    children = [
+        root for root in store.blocks.keys()
+        if store.blocks[root].parent_root == block_root
+    ]
+
+    # If any children branches contain expected finalized/justified checkpoints,
+    # add to filtered block-tree and signal viability to parent.
+    if any(children):
+        if True in [filter_block_tree(store, child, blocks) for child in children]:
+            blocks[block_root] = block
+            return True
+        return False
+
+    # If leaf block, check finalized/justified checkpoints as matching latest.
+    # If matching, add to viable block-tree and signal viability to parent.
+    head_state = store.block_states[block_root]
+
+    # Handle base case where justified hasn't updated yet
+    if head_state.current_justified_checkpoint.epoch == GENESIS_EPOCH:
+        is_viable_branch = True
+    else:
+        is_viable_branch = (
```

Comment:
All this forkchoice PR really does is to limit the tree to blocks that have a finalized checkpoint equal to that of the store? So branches that have a later finalized epoch state, but still descendant of the finalized epoch state in the store, are not considered as head, and their complete branch is ignored? And thus it fails to sync new valid finalized branches? Or is the store finalized checkpoint assumed to always be as up to date with store states as can be? Can confirm that 130 and 144 in the example sketch you sent won't be part of the subtree though in the example as-is.

can we make this check if `head_state.finalized_checkpoint` is equal to, or descendant of, store.finalized_checkpoint instead? And same thing for justified?

Reply:
Because we (potentially) update latest justified and finalized when adding any block to the block tree via `on_block`, the `store.finalized_checkpoint`/`store.justified_checkpoint` is always the highest possible value found in the block tree.

So I'm making the claim that there should never be a branch with a higher latest finalized/justified than that tracked in the store. If this is _not_ the case, then we would want to change it to be as you described.